### PR TITLE
feature/move-to-pf-dal

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Move to use pf::dal and SQL::Abstract::More for database access

Impacts
=======
Database access and sql generation

NEW Package(s) required
=======================
A patched version of SQL::Abstract::More (I submitted the patch to the project but they have not updated CPAN as of yet)
perl-SQL-Abstract-Plugin-InsertMulti

Delete branch after merge
=========================
NO

TODO
=========================
 - [x] Move node search to the dal
 - [x] Have search and find use the same columns

### CODE REVIEW

**Comment by julsemaan** [comment]
Seems the node search wasn't changed in this PR so it goes through DAL. That means it will need to be adjusted for a tenant manually ? Or was that just forgotten ?

**Comment by julsemaan** [comment]
As jazzed, search should default to the same columns as find to ease the usage

**Comment by jrouzierinverse** [comment]
Node search was missed

**Comment by jrouzierinverse** [comment]
search uses the same columns as find now

**Comment by julsemaan** [comment]
This is still missing node search via the DAL according to the todo of the PR

File: lib/pf/Report.pm [hunk 1]
```diff
@@ -37,15 +38,13 @@ has 'person_fields', (is => 'rw', isa => 'ArrayRef[Str]');
...
-# empty since no queries are prepared upfront
-sub Report_db_prepare {}
-
 sub generate_sql_query {
     my ($self, %infos) = @_;
     my $logger = get_logger;
 
     my $sqla = SQL::Abstract::More->new();
     my $and = [];
+    $infos{search}{type} //= $self->base_conditions_operator;
```
Comment (julsemaan) [inline lib/pf/Report.pm:47]: Any specific reason for adding this? Doesn't seem related to DAL work

File: lib/pf/activation.pm [hunk 2]
```diff
@@ -248,13 +236,11 @@ update the status of a given pending activation record
...
     my ($status, $rows) = pf::dal::activation->update_items(
-        {
-            -set => {
-                status => $new_status,,
-            },
-            -where => {
-                code_id => $code_id,
-            }
+        -set => {
+            status => $new_status,,
```
Comment (julsemaan) [inline lib/pf/activation.pm]: Seems like there a useless comma **#**:nail_care:

File: lib/pf/activation.pm [hunk 3]
```diff
@@ -591,17 +571,15 @@ sub set_status_verified_by_mac {
...
-            -columns => [\1],
-            -limit => 1,
-        }
+        -where => {
+            mac => $mac,
+            type => $type,
+            status => $UNVERIFIED,
+            expiration => { ">=" => \['NOW()']},
+        },
+        -from => pf::dal::activation->find_from_tables(),
```
Comment (julsemaan) [inline lib/pf/activation.pm]: Not 100% sure, but I believe we said this shouldn't be used anymore because it should already do this behavior by default

File: lib/pf/violation.pm [hunk 4]
```diff
@@ -178,16 +179,14 @@ sub violation_grace {
     my ( $mac, $vid ) = @_;
 #        qq [ select unix_timestamp(start_date)+grace_period-unix_timestamp(now()) from violation v left join class c on v.vid=c.vid where mac=? and v.vid=? and status="closed" order by start_date desc ]);
```
Comment (julsemaan) [inline lib/pf/violation.pm]: not really useful anymore

File: lib/pf/violation.pm [hunk 5]
```diff
@@ -197,11 +196,9 @@ sub violation_count {
     my ($mac) = @_;
 #        qq [ select count(*) from violation where mac=? and status="open" ]);
```
Comment (julsemaan) [inline lib/pf/violation.pm]: Not really useful anymore

File: lib/pf/violation.pm [hunk 6]
```diff
@@ -212,15 +209,13 @@ sub violation_count_reevaluate_access {
     my ($mac) = @_;
 #       qq [ select count(*) from violation, action where violation.vid=action.vid and action.action='reevaluate_access' and mac=? and status="open" ]);
```
Comment (julsemaan) [inline lib/pf/violation.pm]: Not really useful anymore

File: lib/pf/violation.pm [hunk 7]
```diff
@@ -230,11 +225,9 @@ sub violation_count_vid {
     my ( $mac, $vid ) = @_;
 #        qq [ select count(*) from violation where mac=? and vid=? ]);
```
Comment (julsemaan) [inline lib/pf/violation.pm]: Not really useful anymore

File: lib/pf/violation.pm [hunk 8]
```diff
@@ -245,12 +238,10 @@ sub violation_count_open_vid {
     my ( $mac, $vid ) = @_;
 #        qq [ select count(*) from violation where mac=? and vid=? and status="open" ]);
```
Comment (julsemaan) [inline lib/pf/violation.pm]: Not really useful anymore

File: lib/pf/violation.pm [hunk 9]
```diff
@@ -267,6 +258,7 @@ sub violation_exist {
         },
         -columns => [qw(id mac vid start_date release_date status ticket_ref notes)],
         -limit => 1,
+        -from => pf::dal::violation->table,
```
Comment (julsemaan) [inline lib/pf/violation.pm]: useful ?

File: lib/pf/violation.pm [hunk 10]
```diff
@@ -278,6 +270,7 @@ sub violation_exist_id {
...
```
Comment (julsemaan) [inline lib/pf/violation.pm]: useful ?

File: lib/pf/violation.pm [hunk 11]
```diff
@@ -290,6 +283,7 @@ sub violation_exist_open {
             status => 'open',
         },
         -columns => [qw(id mac vid start_date release_date status ticket_ref notes)],
+        -from => pf::dal::violation->table,
```
Comment (julsemaan) [inline lib/pf/violation.pm]: useful ?

File: lib/pf/violation.pm [hunk 12]
```diff
@@ -330,7 +324,8 @@ sub violation_view_open {
             mac => $mac,
         },
         -columns => [qw(id mac vid start_date release_date status ticket_ref notes)],
-        -order_by => '-start_date',
+        -order_by => { -desc => 'start_date' },
+        -from => pf::dal::violation->table,
```
Comment (julsemaan) [inline lib/pf/violation.pm]: useful ?

File: lib/pf/violation.pm [hunk 13]
```diff
@@ -814,6 +805,7 @@ sub violation_exist_acct {
                 "<=" => \'NOW()',
             },
         },
+        -from => pf::dal::violation->table,
```
Comment (julsemaan) [inline lib/pf/violation.pm]: useful ?

File: lib/pf/violation.pm [hunk 14]
```diff
@@ -832,8 +824,9 @@ sub violation_view_last_closed {
             vid => $vid,
             status => "closed",
         },
-        -order_by => '-release_date',
+        -order_by => {-desc => 'release_date'} ,
         -columns => [qw(mac vid release_date)],
+        -from => pf::dal::violation->table,
```
Comment (julsemaan) [inline lib/pf/violation.pm]: useful ?

File: lib/pf/violation.pm [hunk 15]
```diff
@@ -894,15 +887,17 @@ sub violation_maintenance {
...
-               -limit => $batch,
-               -columns => [qw(id mac vid notes status)],
-            }
+            -where => {
+                status => ["open", "delayed"],
+                release_date => [-and => {"!=" => $ZERO_DATE}, {"<=" => \'NOW()'}],
+            },
+           -limit => $batch,
+           -columns => [qw(id mac vid notes status)],
+           -from => pf::dal::violation->table,
```
Comment (julsemaan) [inline lib/pf/violation.pm]: useful ?

File: lib/pf/Report.pm (hunk 1 above)
Comment (jrouzierinverse) [reply lib/pf/Report.pm:47]: Yes there a undef bug when I was refactoring.
I will added add that to devel.

File: lib/pf/violation.pm (hunk 15 above)
Comment (jrouzierinverse) [reply lib/pf/violation.pm]: Yes it is I did not want to hard code the table name

Comment (julsemaan) [reply lib/pf/violation.pm]: What I mean is that shouldn't this already be the case when using the violation DAL object ?

Comment (julsemaan) [reply lib/pf/violation.pm]: As jazzed verbally, if joins should be skipped, it should be done via an explicit parameter instead of this workaround so its more explicit to the untrained eye

File: lib/pf/violation.pm [hunk 16]
```diff
@@ -258,7 +253,7 @@ sub violation_exist {
         },
         -columns => [qw(id mac vid start_date release_date status ticket_ref notes)],
         -limit => 1,
-        -from => pf::dal::violation->table,
+        -no_default_join => 1,
```
Comment (julsemaan) [inline lib/pf/violation.pm]: use $TRUE constant for this and the ones below

**Comment by jrouzierinverse** [comment]
Finished pf::accounting

**Comment by julsemaan** [comment]
Alright, not sure I'll be able to keep the little sanity I have remaining if I go through and read all of this code again.

Could you rebase this PR and we'll go ahead and merge?

**Comment by extrafu** [comment]
👍 Good job guys.
